File: src/integrations/triadHarness.ts
/**
 * Shared MCOP triad harness used by every ecosystem integration shim
 * (`langchain.ts`, `llamaIndex.ts`, `haystack.ts`).
 *
 * The shims accept either a fully wired triad (encoder + stigmergy + etch),
 * or — most commonly — `undefined`, in which case the harness lazily
 * constructs a single deterministic triad with the project-default
 * configuration. Every shim shares the same harness instance so that
 * traces recorded by `MCOPLangChainMemory` resonate against queries from
 * `MCOPLlamaIndexVectorStore` and etches written by either are visible to
 * `MCOPHaystackMemoryStore` — i.e., the integrations cooperate on a single
 * MCOP memory rather than each being a silo.
 *
 * Every public surface returns a `MCOPProvenance` block carrying the
 * SHA-256 etch hash, the Stigmergy Merkle root, a UUID-v4 traceId, and an
 * ISO8601 timestamp. This is the same provenance shape the deterministic
 * benchmark snapshot guarantees, so callers can stitch ecosystem traffic
 * into the same Holographic Etch ledger that backs the v2.4 reproducible
 * benchmark badge.
 */

import { randomUUID } from 'node:crypto';
import {
  ContextTensor,
  EtchRecord,
  HolographicEtch,
  NovaNeoEncoder,
  PheromoneTrace,
  StigmergyV5,
} from '../core';

/** Triad bundle every ecosystem integration shim is parameterised by. */
export interface MCOPTriad {
  readonly encoder: NovaNeoEncoder;
  readonly stigmergy: StigmergyV5;
  readonly etch: HolographicEtch;
}

/** Optional triad knobs every shim's constructor accepts. */
export interface MCOPTriadOptions {
  readonly triad?: MCOPTriad;
  readonly encoderDimensions?: number;
  readonly resonanceThreshold?: number;
  readonly maxTraces?: number;
  /**
   * Etch confidence floor. Defaults to 0 for the integration shims (every
   * recorded ecosystem event etches) — the dialectical-synthesis path uses
   * a stricter floor to filter low-resonance refinements.
   */
  readonly etchConfidenceFloor?: number;
}

/** Cryptographic provenance attached to every shim response. */
export interface MCOPProvenance {
  readonly traceId: string;
  readonly etchHash: string;
  readonly merkleRoot: string | undefined;
  readonly timestamp: string;
  readonly auditable: boolean;
}

/** Result of recording a single ecosystem event into the triad. */
export interface MCOPRecordResult {
  readonly trace: PheromoneTrace;
  readonly etch: EtchRecord;
  readonly provenance: MCOPProvenance;
}

/** Lazily build (and cache) a default triad if the caller didn't supply one. */
export function ensureTriad(options: MCOPTriadOptions = {}): MCOPTriad {
  if (options.triad) return options.triad;
  const encoder = new NovaNeoEncoder({
    dimensions: options.encoderDimensions ?? 64,
  });
  const stigmergy = new StigmergyV5({
    resonanceThreshold: options.resonanceThreshold ?? 0.55,
    maxTraces: options.maxTraces ?? 2048,
  });
  const etch = new HolographicEtch({
    confidenceFloor: options.etchConfidenceFloor ?? 0,
  });
  return { encoder, stigmergy, etch };
}

/** Encode → record trace → etch a single (text, metadata) pair. */
export function recordIntoTriad(
  triad: MCOPTriad,
  text: string,
  metadata?: Record<string, unknown>,
  note?: string,
): MCOPRecordResult {
  const context = triad.encoder.encode(text);
  const synthesis = synthesiseFromContext(context);
  const trace = triad.stigmergy.recordTrace(context, synthesis, metadata);
  const etch = triad.etch.applyEtch(context, synthesis, note ?? 'mcop-integration-shim');
  const provenance: MCOPProvenance = {
    traceId: randomUUID(),
    etchHash: etch.hash,
    merkleRoot: triad.stigmergy.getMerkleRoot(),
    timestamp: new Date().toISOString(),
    auditable: Boolean(etch.hash) && Boolean(triad.stigmergy.getMerkleRoot()),
  };
  return { trace, etch, provenance };
}

/** Run a resonance query and return the matching trace (if any). */
export function recallFromTriad(
  triad: MCOPTriad,
  query: string,
): { context: ContextTensor; resonance: ReturnType<StigmergyV5['getResonance']> } {
  const context = triad.encoder.encode(query);
  const resonance = triad.stigmergy.getResonance(context);
  return { context, resonance };
}

/**
 * Deterministic synthesis vector derived from the encoded context. We do
 * NOT call out to an LLM here — the integrations are about funnelling
 * ecosystem events into MCOP memory, not about reshaping the triad's
 * synthesis behaviour. Using the context vector itself as the synthesis
 * vector keeps the etch byte-identical-reproducible across runtimes.
 */
function synthesiseFromContext(context: ContextTensor): number[] {
  return context.slice();
}

File: mcop_package/mcop/integrations/triad_harness.py
"""Shared triad harness for the optional ecosystem integration shims.

The historical ``MCOP*`` imports, constructor seeds, and normal method calls
remain compatible, but they now wrap the canonical flagship implementations in
:mod:`mcop.triad`.  This removes a second, drifting Stigmergy/Etch
implementation while retaining the integration defaults used by LangChain,
LlamaIndex, and Haystack adapters.  The bounded ``traces`` and ``etches``
properties return snapshots; mutating those returned lists does not mutate the
internal rings.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Sequence, Tuple

from ..triad import (
    EtchRecord,
    HolographicEtch,
    NovaNeoEncoder,
    PheromoneTrace,
    ResonanceResult,
    StigmergyV5,
)

__all__ = [
    "MCOPEncoder",
    "MCOPEtchRecord",
    "MCOPHolographicEtch",
    "MCOPPheromoneTrace",
    "MCOPProvenance",
    "MCOPRecordResult",
    "MCOPResonanceResult",
    "MCOPStigmergy",
    "MCOPTriad",
    "MCOPTriadOptions",
    "ensure_triad",
    "recall_from_triad",
    "record_into_triad",
]


class MCOPEncoder(NovaNeoEncoder):
    """Compatibility wrapper with the integration default of 64 dimensions."""

    def __init__(
        self, dimensions: int = 64, normalize: bool = False, backend: str = "hash"
    ) -> None:
        super().__init__(dimensions=dimensions, normalize=normalize, backend=backend)


class MCOPStigmergy(StigmergyV5):
    """Compatibility wrapper retaining the integration threshold of 0.55."""

    def __init__(
        self,
        resonance_threshold: float = 0.55,
        max_traces: int = 2048,
        traces: Optional[Sequence[PheromoneTrace]] = None,
        **kwargs: Any,
    ) -> None:
        # The historical integration memory used a fixed threshold. Preserve
        # that behavior unless a caller explicitly opts into adaptation.
        adaptive_threshold = kwargs.pop("adaptive_threshold", False)
        super().__init__(
            resonance_threshold=resonance_threshold,
            max_traces=max_traces,
            adaptive_threshold=adaptive_threshold,
            **kwargs,
        )
        # Legacy dataclass construction accepted a seed list. Seed through the
        # canonical ring so capacity and parent-root invariants remain intact.
        for trace in traces or ():
            self._traces.push(trace)


class MCOPHolographicEtch(HolographicEtch):
    """Compatibility wrapper that etches every integration event by default."""

    def __init__(
        self,
        confidence_floor: float = 0.0,
        etches: Optional[Sequence[EtchRecord]] = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(confidence_floor=confidence_floor, **kwargs)
        for record in etches or ():
            self._etches.push(record)
            if self.audit_log:
                self._audit.push(record)


# Type aliases preserve all historical import paths while records are now
# produced by the one canonical implementation.
MCOPPheromoneTrace = PheromoneTrace
MCOPResonanceResult = ResonanceResult
MCOPEtchRecord = EtchRecord


@dataclass(frozen=True)
class MCOPProvenance:
    trace_id: str
    etch_hash: str
    merkle_root: Optional[str]
    timestamp: str
    auditable: bool


@dataclass(frozen=True)
class MCOPRecordResult:
    trace: PheromoneTrace
    etch: EtchRecord
    provenance: MCOPProvenance


@dataclass
class MCOPTriad:
    encoder: MCOPEncoder
    stigmergy: MCOPStigmergy
    etch: MCOPHolographicEtch


@dataclass(frozen=True)
class MCOPTriadOptions:
    triad: Optional[MCOPTriad] = None
    encoder_dimensions: int = 64
    resonance_threshold: float = 0.55
    max_traces: int = 2048
    etch_confidence_floor: float = 0.0


def ensure_triad(options: Optional[MCOPTriadOptions] = None) -> MCOPTriad:
    """Build a deterministic triad, or return the supplied bundle verbatim."""
    selected = options or MCOPTriadOptions()
    if selected.triad is not None:
        return selected.triad
    return MCOPTriad(
        encoder=MCOPEncoder(dimensions=selected.encoder_dimensions),
        stigmergy=MCOPStigmergy(
            resonance_threshold=selected.resonance_threshold,
            max_traces=selected.max_traces,
        ),
        etch=MCOPHolographicEtch(
            confidence_floor=selected.etch_confidence_floor
        ),
    )


def record_into_triad(
    triad: MCOPTriad,
    text: str,
    metadata: Optional[Dict[str, Any]] = None,
    note: Optional[str] = None,
) -> MCOPRecordResult:
    """Encode, record, and etch one ecosystem event."""
    context = triad.encoder.encode(text)
    synthesis = list(context)
    trace = triad.stigmergy.record_trace(context, synthesis, metadata)
    etch = triad.etch.apply_etch(
        context,
        synthesis,
        note="mcop-integration-shim" if note is None else note,
    )
    root = triad.stigmergy.get_merkle_root()
    provenance = MCOPProvenance(
        trace_id=trace.id,
        etch_hash=etch.hash,
        merkle_root=root,
        timestamp=_iso_utc_now(),
        auditable=bool(etch.hash) and root is not None,
    )
    return MCOPRecordResult(trace=trace, etch=etch, provenance=provenance)


def recall_from_triad(
    triad: MCOPTriad, query: str
) -> Tuple[List[float], ResonanceResult]:
    """Encode a query and return ``(context, resonance)``."""
    context = triad.encoder.encode(query)
    return context, triad.stigmergy.get_resonance(context)


def _iso_utc_now() -> str:
    return datetime.now(tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
